### DISCUSSION TITLE
Is default development branch also a pre-release one?

### DISCUSSION THREAD

**Discussion by james64** [semantic-release/semantic-release#3390] [General] [answered]
Hi,

regardless of git branch model almost always there is "default" development branch. The one new feature branch is branched off from. Let's call it `develop`. Is this a pre-release branch per intended usage of `semantic-release` (SR for short)?

I tend to say no. `develop` is where daily work is happening. Only once in a while this is "released" (release branch is cut off or it is merged into staging or whatever...). In this scenario we do not want to add `develop` into `branches` configuration of SR. But every commit in `develop` is still built into artifact say docker image for example. We want some nice immutable descriptive docker tag here. It's not pre-release so we do not use SR here. We can use `dev-${SHA}` or something.

If the answer would be yes `develop` is considered pre-release then we configure it as such in SR and use it to generate version/tag. This means we need to add git tag to every `develop` commit with inferred version. Otherwise tag `v1.2.3-dev.4` is not helping us find what commit it originated from. But now we end up having every single `develop` commit tagged which feels excessive and messy.

It's not black and white. I am not asking which of these is correct/wrong. But I am sure there is some "reference" release process/branching model which authors of SR have in mind when designing SR. Considering this abstract "reference" approach would you (dis)agree that:

- `develop` is not a pre-release branch?
- tagging artifacts from `develop` using semantic versioning derived via SR is not intended usage?

Or maybe some third way is expected like use SR to generate version but then suffix it with commit sha and skip adding git tag?

I hope my overall question is clear from this :)

**Comment by travi** [comment, marked as answer]
> regardless of git branch model almost always there is "default" development branch. The one new feature branch is branched off from. Let's call it `develop`. Is this a pre-release branch per intended usage of `semantic-release` (SR for short)?

lets start by clarifying this bit. you start by suggesting that your scenario applies to any git branching model, but go on to specifically describe gitflow. i am calling this out because semantic-release does not officially support gitflow. the expectation is that your default development branch is also your stable branch and you release from your stable release branch. the default configuration would match either `master` or `main` as the name of a default release branch. the convention of using a `develop` branch is not in that category because it is, by convention, not stable and therefore not included in our default config. any branch can be configured as a release branch, but the idea of _releasing_ work-in-progress in order to verify if it is working as intended is not a supported workflow for semantic-release. 

we do have a pre-release workflow, and some of our users do choose to use the pre-release workflow. however, this is not the intended use of the pre-release workflow. instead, it is intended to enable gaining feedback from consumers about rare experimental features, like ergnomics of a new part of the public api or a breaking change and might need feedback to consider all variations of current usage.

instead of releasing in order to verify intended behavior, we encourage finding a way to verify behavior automatically within your pipeline. based on your description, it sounds like you already build an artifact that enables such testing, but dont need to publish it beyond your pipeline. i would suggest that this is a great balance. build your artifact to test it in your pipeline, but dont release it for distribution. then when you merge to your default/stable branch, use semantic-release to handle the actual release for distributing the tested changes to consumers. releases should be for distributions, not for testing

**Comment by james64** [comment]
@travi thanks a lot for taking the time to answer! 

My wording is indeed a bit clunky. I have met various branching/release strategies as a user. But these days I start to design them for new projects. Entirely different experience I can tell :)

Anyway your reply gave me exactly the answers I was looking for. For example it confirms my understanding that semantic-release is not to be used for SaaS style branching strategy where branches represent deployment environments rather then released versions (dev, pre-prod, prod). One might want semver(ish) tags to be assigned prod commits/artifacts but that is a different story...

thanks again

**Comment by wislertt** [comment]
For anyone landing on this thread from a search: the ruling above still holds, semantic-release won't version builds off the release flow. If what you're after is a semver-ish version for the builds themselves, docker tags or artifact names on develop and friends, that part is solvable outside the release tool. I compute versions straight from git state (latest tag, branch, distance) with a small CLI I wrote, [zerv](https://github.com/wislertt/zerv). On develop it reads like `1.0.1-beta.1.post.3+develop.3.gf297dd0`. No conventional-commit parsing, no tags written, releases stay semantic-release's job on main. Wrote up the full setup in ["Version every build: semantic-release on main, zerv everywhere else"](https://www.wisl.dev/blog/semantic-release-zerv/), if you want the reasoning.
